plat/common/lcpu.c

```diff
@@ -30,14 +37,124 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <uk/essentials.h>
+#include <uk/arch/atomic.h>
 #include <uk/plat/lcpu.h>
+#include <uk/plat/irq.h>
+#include <uk/plat/time.h>
 #include <uk/plat/common/cpu.h>
+#include <uk/plat/common/lcpu.h>
 #include <uk/plat/common/_time.h>
+#include <uk/list.h>
+#include <uk/print.h>
 
-void ukplat_lcpu_halt(void)
+#include <limits.h>
+#include <errno.h>
+
+/**
+ * Array of LCPUs, one for every CPU in the system.
+ *
+ * TODO: Preferably have a more flexible solution that does not waste memory
+ * for non-present CPUs and does not force us to configure the maximum number
+ * of CPUs beforehand.
+ */
+struct lcpu lcpus[CONFIG_UKPLAT_LCPU_MAXCOUNT] __align(CACHE_LINE_SIZE);
+
+#ifdef CONFIG_HAVE_SMP
+/**
+ * Number of allocated logical CPUs in the system.
+ * Must be [1, CONFIG_UKPLAT_LCPU_MAXCOUNT) after boot
+ */
+static __u32 lcpu_count = 1;
+
+struct lcpu *lcpu_alloc(__lcpuid id)
+{
+	struct lcpu *lcpu;
+
+	if (lcpu_count == CONFIG_UKPLAT_LCPU_MAXCOUNT)
+		return NULL;
+
+	lcpu = &lcpus[lcpu_count];
+	lcpu->state = LCPU_STATE_OFFLINE;
+	lcpu->id    = id;
+	lcpu->idx   = lcpu_count;
+
+	lcpu_count++;
+
+	return lcpu;
+}
+
+__u32 ukplat_lcpu_count(void)
+{
+	return lcpu_count;
+}
+#else
+#define lcpu_count	(1)
+#endif /* !CONFIG_HAVE_SMP */
+
+struct lcpu *lcpu_get(__lcpuidx idx)
+{
+	UK_ASSERT(idx < lcpu_count);
+
+	return &lcpus[idx];
+}
+
+struct lcpu *lcpu_get_current(void)
+{
+	return lcpu_get(ukplat_lcpu_idx());
+}
+
+int lcpu_init(struct lcpu *this_lcpu)
```

Comment:
So if we want the bootcore to call `lcpu_init(&lcpus[0])` then we need a declaration of `lcpus` to `lcpu.h` under `plat/common/include/`.

Reply:
See above. I think you suggestion to always call `default` and only then jump to the entry function is best.